### PR TITLE
[WIP] Numpy kron operator

### PR DESCRIPTION
**Pull request by ckt624**
## Description ##
https://docs.scipy.org/doc/numpy/reference/generated/numpy.kron.html#numpy.kron
Also change check_numerical_grad for 0 size and 0 dim inputs.
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Implements np.kron

## Comments ##
@haojin2 @reminisce

### CODE REVIEW

File: src/operator/numpy/np_kron.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+  }
+
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, out_shape);
+
+  return shape_is_known(*in_attrs) && shape_is_known(*out_attrs);
+}
+
+NNVM_REGISTER_OP(_np_kron)
+.add_alias("_npi_kron")
```
Comment (haojin2) [inline src/operator/numpy/np_kron.cc]: Do you want to expose the operator directly to users or not?
If no then just do
```c++
NNVM_REGISTER_OP(_npi_kron)
```

File: tests/python/unittest/test_numpy_op.py [hunk 2]
```diff
@@ -829,6 +829,57 @@ def get_indices(axis_size):
...
+                assert_almost_equal(mx_out.asnumpy(), np_out, rtol=1e-3, atol=1e-5)
+
+                # test numeric gradient
+                if (len(a_shape) > 0 and len(b_shape) > 0 and _np.prod(a_shape) > 0 
+                  and _np.prod(b_shape) > 0):
+                    a_sym = mx.sym.Variable("a").as_np_ndarray()
+                    b_sym = mx.sym.Variable("b").as_np_ndarray()
+                    mx_sym = mx.sym.np.kron(a_sym, b_sym).as_nd_ndarray()
+                    check_numeric_gradient(mx_sym, [a.as_nd_ndarray(), b.as_nd_ndarray()],
+                      rtol=1e-1, atol=1e-1, dtype=dtype)
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: Have you tried smaller numbers for `atol` and `rtol` ? Something like `rtol=1e-2` and `atol=1e-4`?

File: src/operator/numpy/np_kron.cu [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file np_kron.cu
+ * \brief GPU Implementation of numpy-compatible kron
```
Comment (haojin2) [inline src/operator/numpy/np_kron.cu]: "numpy-compatible Kronecker product"

File: src/operator/numpy/np_kron-inl.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  using namespace mxnet_op;
+
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+
+  const TBlob& a = inputs[0];
+  const TBlob& b = inputs[1];
+  const TBlob& out = outputs[0];
+  const mxnet::TShape a_shape = a.shape_;
+  const mxnet::TShape b_shape = b.shape_;
```
Comment (haojin2) [inline src/operator/numpy/np_kron-inl.h]: if `a_shape` and `b_shape` not used at all then you should remove the 2 lines.

File: src/operator/numpy/np_kron-inl.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 2U);
+
+  const TBlob& ograd = inputs[0];
+  const TBlob& a = inputs[1];
+  const TBlob& b = inputs[2];
+  const TBlob& grad_a = outputs[0];
+  const TBlob& grad_b = outputs[1];
+  const mxnet::TShape a_shape = a.shape_;
+  const mxnet::TShape b_shape = b.shape_;
```
Comment (haojin2) [inline src/operator/numpy/np_kron-inl.h]: Same here

File: src/operator/numpy/np_kron-inl.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+template<typename xpu>
+inline void KronOpBackward(const nnvm::NodeAttrs& attrs,
+                           const OpContext& ctx,
+                           const std::vector<TBlob>& inputs,
+                           const std::vector<OpReqType>& req,
+                           const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow_op;
```
Comment (haojin2) [inline src/operator/numpy/np_kron-inl.h]: Seems like nothing in this function is using something from `mshadow_op` namespace, may remove this line?

File: src/operator/numpy/np_kron-inl.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+namespace op {
+
+template<typename xpu>
+inline void KronOpForward(const nnvm::NodeAttrs& attrs,
+                          const OpContext& ctx,
+                          const std::vector<TBlob>& inputs,
+                          const std::vector<OpReqType>& req,
+                          const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
```
Comment (haojin2) [inline src/operator/numpy/np_kron-inl.h]: Seems like nothing in this function is using something from `mxnet_op` namespace, may remove this line?

File: src/operator/numpy/np_kron-inl.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+#ifndef MXNET_OPERATOR_NUMPY_NP_KRON_INL_H_
+#define MXNET_OPERATOR_NUMPY_NP_KRON_INL_H_
+
+#include <mxnet/operator_util.h>
+#include <vector>
+#include "../tensor/dot-inl.h"
+#include "../tensor/elemwise_binary_op.h"
+#include "../tensor/broadcast_reduce_op.h"
```
Comment (haojin2) [inline src/operator/numpy/np_kron-inl.h]: Some of the includes seem redundant to me, check and remove all redundant includes, same for all other new c++ files.

File: src/operator/numpy/np_kron.cc (hunk 1 above)
Comment (ckt624) [reply src/operator/numpy/np_kron.cc]: Fixed. Thx.

File: tests/python/unittest/test_numpy_op.py (hunk 2 above)
Comment (ckt624) [reply tests/python/unittest/test_numpy_op.py]: Fixed. Thx.

File: src/operator/numpy/np_kron.cu (hunk 3 above)
Comment (ckt624) [reply src/operator/numpy/np_kron.cu]: Fixed. Thx.

File: src/operator/numpy/np_kron-inl.h (hunk 4 above)
Comment (ckt624) [reply src/operator/numpy/np_kron-inl.h]: Fixed. Thx.

File: src/operator/numpy/np_kron-inl.h (hunk 5 above)
Comment (ckt624) [reply src/operator/numpy/np_kron-inl.h]: Fixed. Thx.

File: src/operator/numpy/np_kron-inl.h (hunk 6 above)
Comment (ckt624) [reply src/operator/numpy/np_kron-inl.h]: Fixed. Thx.

File: src/operator/numpy/np_kron-inl.h (hunk 7 above)
Comment (ckt624) [reply src/operator/numpy/np_kron-inl.h]: Fixed. Thx.

File: src/operator/numpy/np_kron-inl.h (hunk 8 above)
Comment (ckt624) [reply src/operator/numpy/np_kron-inl.h]: Fixed. Thx.

File: src/operator/numpy/np_kron.cc [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+ */
+
+#include "./np_kron-inl.h"
+
+namespace mxnet {
+namespace op {
+
+inline bool KronOpShape(const nnvm::NodeAttrs& attrs,
+                          mxnet::ShapeVector *in_attrs,
+                          mxnet::ShapeVector *out_attrs) {
```
Comment (haojin2) [inline src/operator/numpy/np_kron.cc]: Align the arguments.

Comment (ckt624) [reply src/operator/numpy/np_kron.cc]: fixed. Thx.

**Comment by reminisce** [comment]
This PR only supports the second argument with `ndim <= 1`. Need more implementation work to make it complete.

**Pull request closed without merging**